p2p/chain.py

```diff
@@ -152,61 +156,92 @@ def register_peer(self, peer: BasePeer) -> None:
                 await self.chaindb.coro_persist_header(header)
                 start_at = header.block_number
 
-            # TODO: Queue receipt downloads.
             self._body_requests.put_nowait(headers)
+            self._receipt_requests.put_nowait(headers)
 
             self.logger.debug("Asking %s for header batch starting at %d", peer, start_at)
             # TODO: Instead of requesting sequential batches from a single peer, request a header
             # skeleton and make concurrent requests, using as many peers as possible, to fill the
             # skeleton.
             peer.sub_proto.send_get_block_headers(start_at, eth.MAX_HEADERS_FETCH, reverse=False)
 
-    async def body_downloader(self) -> None:
+    async def _downloader(self,
+                          queue: 'asyncio.Queue[List[BlockHeader]]',
+                          should_skip: Callable[[BlockHeader], bool],
+                          request_func: Callable[[List[BlockHeader]], Awaitable[None]],
+                          pending: Dict[Any, Tuple[BlockHeader, float]]) -> None:
         while True:
             try:
                 headers = await wait_with_token(
-                    self._body_requests.get(),
+                    queue.get(),
                     token=self.cancel_token,
                     timeout=self._reply_timeout)
             except TimeoutError:
-                # We use a timeout above just to make sure we periodically retry timedout body
-                # requests even when there's no new bodies coming through.
+                # We use a timeout above to make sure we periodically retry timedout items
+                # even when there's no new items coming through.
                 headers = []
             except OperationCancelled:
                 return
 
             if headers:
-                headers_with_non_empty_bodies = [
-                    header for header in headers
-                    if (header.transaction_root != self.chaindb.empty_root_hash or
-                        header.uncles_hash != EMPTY_UNCLE_HASH)
-                ]
-                self.request_bodies(headers_with_non_empty_bodies)
+                await request_func(
```

Comment:
If this moved into an `else` block from the `try/except` above, you could remove the `headers = []` from inside of the `TimeoutError` block.